src/Rings/MPolyQuo.jl

```diff
@@ -101,6 +101,64 @@ function Base.deepcopy_internal(a::MPolyQuoRingElem, dict::IdDict)
   return MPolyQuoRingElem(Base.deepcopy_internal(a.f, dict), a.P, a.simplified)
 end
 
+########################################################################
+# Representatives of elements in quotient rings and normal forms 
+#
+# Elements [a] ∈ P/I in quotients of polynomial rings 
+# P = 𝕜[x₁,…,xₙ] by ideals I = ⟨f₁,…,fᵣ⟩ admit unique representatives 
+# a ∈ P whenever a normal form algorithm exists for the polynomial 
+# ring P. This is really a question about the ring of coefficients 𝕜. 
+#
+# In general, we can not expect a normal form algorithm to exist and, 
+# in particular, that it is implemented in Singular. However, we wish 
+# to use Singular as a default backend and this also drives the design 
+# of the quotient rings to begin with. 
+#
+# To make sure that the data structure for quotient rings can 
+# nevertheless also accomodate more exotic coefficient rings we 
+# provide the following functionality to decide the existence and use 
+# of a singular backend depending on the type.
+########################################################################
+
+# The trait of the coefficient ring and its elements to decide on 
+# whether or not we do have a singular backend for groebner basis 
+# computations
+abstract type HasSingularGroebnerAlgorithmTrait end
+abstract type HasSingularNormalFormTrait <: HasSingularGroebnerAlgorithmTrait end
+
+# A normal form algorithm (for possibly non-global orderings) requires 
+# strictly more, so we make it a subcase.
+struct HasSingularNormalForm <: HasSingularNormalFormTrait end
+struct HasNoSingularNormalForm <: HasSingularNormalFormTrait end
+
+struct HasSingularGroebnerAlgorithm <: HasSingularGroebnerAlgorithmTrait end
+struct HasNoSingularGroebnerAlgorithm <: HasSingularGroebnerAlgorithmTrait end
+
+# By default we do not expect the Singular backend to handle the case
+HasSingularNormalFormTrait(a::Any) = HasSingularNormalFormTrait(typeof(a))
+HasSingularNormalFormTrait(::Type{T}) where {T} = HasNoSingularNormalForm()
```

Comment:
```suggestion
HasSingularNormalFormTrait(R::NCRing) = HasSingularNormalFormTrait(typeof(a))
HasSingularNormalFormTrait(a::NCRingElem) = HasSingularNormalFormTrait(parent(a))
HasSingularNormalFormTrait(::Type{T}) where {T} = HasNoSingularNormalForm()
```
Wouldn't this make it a bit easier, as down below you would only need to specify traits for the container types, and not for element types? This seems like a good place for future errors (having differences between traits of containers and its elements).

Reply:
Alright, this seems reasonable.

Reply:
```suggestion
HasSingularNormalFormTrait(a::Any) = HasSingularNormalFormTrait(typeof(a))
HasSingularNormalFormTrait(::Type{T}) where {T <: NCRingElem} = HasSingularNormalFormTrait(parent_type(T))
HasSingularNormalFormTrait(::Type{T}) where {T <: NCRing} = HasNoSingularNormalForm()
```
Or even like this. Then, it works for all four of container, container type, element, element type. While one still only has to define it for the container type (as in https://github.com/oscar-system/Oscar.jl/pull/2789#discussion_r1324235900).